Feature/multicast sdk

ShepelIlya opened the pull request:
Added multicast data receiver. To use this you need to set **udp_dest** on ouster configuration page to something like "239.xxx.xxx.xxx" and also set parameter **mtp_dest** in launch file to IP address of host interface where you want to receive multicast data (for adding this interface to multicast group). 

Tested on three Jetson Xavier and two Ouster OS-1 Rev D with FW 2.4

[Link](https://github.com/ouster-lidar/ouster-ros/pull/57) to corresponding PR in ros driver

Samahu commented on ouster_client/src/client.cpp:
```diff
@@ -380,6 +464,55 @@ std::shared_ptr<client> init_client(const std::string& hostname,
+    cli->hostname = hostname;
+
+    cli->lidar_fd = mtp_data_socket(lidar_port, udp_dest_host, mtp_dest_host);
+    cli->imu_fd = mtp_data_socket(imu_port); // no need to join multicast group
```
May I ask why not be that case?

Samahu replied:
Because of this I start the 2nd client I don't get IMU messages, I also tried to change the line to `cli->imu_fd = mtp_data_socket(imu_port, udp_dest_host, mtp_dest_host);` yet I still didn't get imu packets on the 2nd client, any idea why?

ShepelIlya replied:
I can check it and try to fix it tomorrow

ShepelIlya replied:
Can you check it again? Cause i can see imu packets on both clients now, main and slave. May be you configuration is wrong? 
I check it at this [commit of sdk](https://github.com/ouster-lidar/ouster_example/pull/480/commits/cf2892afbd626ec057880a119d27029753511203) and [commit of driver](https://github.com/ouster-lidar/ouster-ros/pull/57/commits/c6d8d4e4eebfb4ad60e19973dafd6b2423b535d5).

My launch files for master and slave are attached (renamed to txt)
[sensor_mtp_main.launch.txt](https://github.com/ouster-lidar/ouster_example/files/10732570/sensor_mtp_main.launch.txt)
[sensor_mtp_slave.launch.txt](https://github.com/ouster-lidar/ouster_example/files/10732571/sensor_mtp_slave.launch.txt)

By the way - we dont need to join to multicast group, because we already in it after joining in function for lidar packets. No need to do it twice.

Samahu commented on ouster_client/include/ouster/types.h:
```diff
@@ -225,6 +225,8 @@ using ColumnWindow = std::pair<int, int>;
 struct sensor_config {
     optional<std::string> udp_dest;  ///< The destination address for the
                                      ///< lidar/imu data to be sent to
+    optional<std::string> mtp_dest;  ///< The host IP address of interfece for the
+                                     ///< lidar/imu data to be sent to via multicast
```
You should be able to make the necessary changes, I will try to point out the necessary changes while going through the review

Samahu commented on ouster_client/src/client.cpp:
```diff
@@ -380,6 +464,55 @@ std::shared_ptr<client> init_client(const std::string& hostname,
     return cli;
 }
 
+std::shared_ptr<client> init_client(const std::string& hostname,                                    
+                                    const std::string& udp_dest_host,
+                                    const std::string& mtp_dest_host,
+                                    lidar_mode ld_mode, timestamp_mode ts_mode,
+                                    int lidar_port, int imu_port,
+                                    int timeout_sec) {
```
This works but I do have concerns where a user invokes it from one client which initializes the sensor with certain settings, on another client the user may accidentally change few settings (like ports, ...) which in the very least will interrupt the stream on the first client due to sensor re-init. On the other hand it could probably break the first client if the user selects a different lidar mode for example. I would imaging this would cause a fault on the first client.

We can do reduce the chance of improper use of the API breaking the call into two separate methods (for example `mtp_init_client_main` and `mtp_init_client_slave` or `_secondary`), first method has the option to modify config while the other can't.

On the other hand, we can still use one method but perform due diligence within the provided roslaunch files to protect against improper use. I am leaning towards breaking this API into two separate APIs and cascade the changes to the launch files too so there is no room for misuse. Something like this:

```suggestion
std::shared_ptr<client> mtp_init_client_main(const std::string& hostname,                                    
                                    const sensor_config& config,
                                    const std::string& mtp_dest_host = "",
                                    int timeout_sec) {

std::shared_ptr<client> mtp_init_client_slave(const std::string& hostname,                                    
                                    const std::string& udp_dest_host,
                                    const std::string& mtp_dest_host,
                                    int lidar_port, int imu_port,
                                    int timeout_sec) {
```

I will weigh more on this on Monday.

ShepelIlya replied:
@Samahu check my last commit, please. I tried to split the mtp_init_client api according to your suggestions

Samahu commented on ouster_client/src/client.cpp:
```diff
@@ -154,6 +154,90 @@ SOCKET udp_data_socket(int port) {
 }
 
+
+SOCKET mtp_data_socket(int port, const std::string& udp_dest_host = "",
```
@Mattc11 I think you are quite familiar multicast, mind take a look at this?

MattC11 replied:
It looks good to me, only comment is the INADD_ANY.

ShepelIlya replied:
Done in [6a42349](https://github.com/ouster-lidar/ouster_example/pull/480/commits/6a42349849740291551c11b631c8424fee9732e3)

Samahu reviewed (commented):
@ShepelIlya Agains thanks you for the contribution, I did test your branch and got to two separate machines/clients connected to the same sensor using multicast.

I had few comments, and few questions. I had few are certain points (like the one against introducing mtp_dest to sensor_config).

With regard to breaking the `init_client(multicast version)` into two separate calls I would like to gather some requirements and get the team consensus on (and get back to you).

ShepelIlya commented:
I understood the logic with the sensor_config structure and moved the mtp_dest parameter to the OusterSensor class.

MattC11 commented on ouster_client/src/client.cpp:
```diff
@@ -154,6 +154,90 @@ SOCKET udp_data_socket(int port) {
+
+            // bind() succeeded; join to multicast group on with preferred address
+            // connect only if addresses are not empty
+            if (!udp_dest_host.empty() && !mtp_dest_host.empty()) {
```
So that people don't need to specify the interface can we default to INADDR_ANY if the mtp_dest_host is not specified. 
Something like:
```
if (!udp_dest_host.empty()) {
                ip_mreq mreq;
                mreq.imr_multiaddr.s_addr = inet_addr(udp_dest_host.c_str());
                if (!mtp_dest_host.empty()) {
                    mreq.imr_interface.s_addr =
                        inet_addr(mtp_dest_host.c_str());
                } else {
                    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
                }
```

Samahu replied:
@MattC11 The second ` if (!mtp_dest_host.empty()) {` will always be true, did you instead mean to say?
```cpp
if (!udp_dest_host.empty()) {
    ip_mreq mreq;
    mreq.imr_multiaddr.s_addr = inet_addr(udp_dest_host.c_str());
    if (!mtp_dest_host.empty()) {
        mreq.imr_interface.s_addr =
        inet_addr(mtp_dest_host.c_str());
    } // <- THIS BRACE WAS PROBABLY MISSING      
} else {
    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
}
```

MattC11 replied:
I think I had it right, udp_dest_host needs to be set. Note that I removed the second part of the if on line 204. That is the multicast ip address that the sensor is sending to so if it doesn't exist we don't have anything to subscribe to. And actually if it isn't set maybe we should exit the whole function right away?

The check I was trying to add is if the mtp_dest_host (this client machine IP address of one of it's interfaces) is empty in which case we subscribe to any interface.

ShepelIlya replied:
But what if we have several physical interfaces? in that case INADDR_ANY will return 1st iface. For example: our case. We have several Vecow EAC-3000 (Jetson Xavier) that have 5 physical interfaces on our robot. One of them we use to collect  sensor data from Ousters (it is a noticeably traffic if we use 2 OS0-64) and another for sending/receiving ROS communication messages for autonomus driving. 

May be i can try to make INADDR_ANY default option for mtp_dest in launch?

ShepelIlya replied:
Yeah, @MattC11 is right - udp_dest_host needs to be set. And working code looks so:
```suggestion
            // bind() succeeded; join to multicast group on with preferred address
            // connect only if addresses are not empty
            if (!udp_dest_host.empty()) {
                ip_mreq mreq;
                mreq.imr_multiaddr.s_addr = inet_addr(udp_dest_host.c_str());
                if (!mtp_dest_host.empty()) {
                    mreq.imr_interface.s_addr = inet_addr(mtp_dest_host.c_str());
                } else {
                    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
                }

                if (setsockopt(sock_fd, IPPROTO_IP, IP_ADD_MEMBERSHIP, (char*)&mreq,
                            sizeof(mreq))) {
                    logger().warn("mtp setsockopt(): {}", impl::socket_get_error());
                    impl::socket_close(sock_fd);
                    continue;
                }
            }
```

But(!), there is small problem: [here](https://github.com/ShepelIlya/ouster-ros/blob/feature/multicast_ros_driver/src/os_sensor_nodelet.cpp#L300-L305) i use mtp_dest_arg to check if we need to use multicast version of init_client.

I can add default constant  INADDR_ANY in launch like you did with possible values of timestamp mode 
OR
I can add additional rosparam to check if we need multicast version.

Which option do you prefer? Or may be smth else?

MattC11 replied:
I'd be inclined to just check if the address range is in the multicast range which will determine which function you call (line 166 of the os_sensor_nodelet.cpp) . A number of ways to do it here: https://stackoverflow.com/questions/22015308/c-function-to-determine-if-ip-address-is-multicast-address
That way you don't need any extra user input.

MattC11 replied:
Alternatively, yes putting it to INADDR_ANY in your launch file so that you have to specify mtp_dest_arg it if you want to be multicast sounds reasonable as well.

ShepelIlya replied:
> I'd be inclined to just check if the address range is in the multicast range which will determine which function you call (line 166 of the os_sensor_nodelet.cpp) . A number of ways to do it here: https://stackoverflow.com/questions/22015308/c-function-to-determine-if-ip-address-is-multicast-address That way you don't need any extra user input.

Yeah, thats good point! I`ll do it

ShepelIlya replied:
I think  [this](https://github.com/ouster-lidar/ouster_example/pull/480/commits/6a42349849740291551c11b631c8424fee9732e3) should work, can you pls check it too with your test code @MattC11 ?

Corresponding [commit](https://github.com/ouster-lidar/ouster-ros/pull/57/commits/200c254f2dff242565a0e1a15e4ea70b5f8eec5e) in ouster_ros.

MattC11 replied:
I tried it out with my test code and it worked.

ShepelIlya replied:
Nice, than i resolve it

MattC11 commented on ouster_client/include/ouster/client.h:
```diff
@@ -174,9 +218,12 @@ bool get_config(const std::string& hostname, sensor_config& config,
+    CONFIG_FORCE_REINIT     = (1 << 2), ///< Forces the sensor to re-init during
                                         ///< set_config even when config params
                                         ///< have not changed
+    CONFIG_GET_ACTIVE       = (1 << 3)  ///< Get active config from sensor  
```
Are you still using this?

Samahu replied:
I didn't observe this before, @ShepelIlya is this really needed?

ShepelIlya replied:
I think that with this we can make initialization of secondary clients more secure. So core idea to read full-config on secondary sensors, without configuring them. And i see 2 ways of doing this - 
1) additional flag that keeps us from configuring
2) change OnInit function, something like that
`class OusterSensor : public OusterClientBase {
   private:
    virtual void onInit() override {
        auto& pnh = getPrivateNodeHandle();
        sensor_hostname = get_sensor_hostname(pnh);
        sensor::sensor_config config;
        u_int8_t flags;
        std::tie(config, flags) = create_sensor_config_rosparams(pnh);
        if (/* maybe we need rosparam here?*/)
             configure_sensor(sensor_hostname, config, flags);
        else
             get_config(sensor_hostname, config, true); 
        sensor_client = create_sensor_client(sensor_hostname, config);
        update_config_and_metadata(*sensor_client);
        save_metadata(pnh);
        OusterClientBase::onInit();
        create_get_config_service();
        create_set_config_service();
        start_connection_loop();
    }`
    
    If you see more correct way to do it  then i can try to implement it.

MattC11 replied:
@Samahu Will have to comment on this one again. I was paying attention to the ros nodelet when I made my comment and he knows a lot more about it.

MattC11 commented on ouster_client/include/ouster/client.h:
```diff
@@ -97,6 +97,50 @@ std::shared_ptr<client> init_client(const std::string& hostname,
+ *
+ * @return pointer owning the resources associated with the connection.
+ */
+std::shared_ptr<client> mtp_init_client_slave(const std::string& hostname,                                    
```
Can you rename to:
mtp_init_client_secondary

ShepelIlya replied:
Done in [77650a5](https://github.com/ouster-lidar/ouster_example/pull/480/commits/77650a5c611b0657ab9b5247802f2d20bc89fbd7)

MattC11 commented:
I've attached a file I used in examples to test this:
`#include <iostream>

#include "ouster/client.h"

using namespace ouster;

const size_t UDP_BUF_SIZE = 65536;

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "\n\nUsage: <sensor_hostname> <main secondary>"
                  << std::endl;

        return argc == 1 ? EXIT_SUCCESS : EXIT_FAILURE;
    }

    const std::string sensor_hostname = argv[1];
    // sensor::sensor_config config;

    bool main = false;

    if (std::string(argv[2]) == "main") {
        main = true;
    } else if (std::string(argv[2]) == "secondary") {
        main = false;
    } else {
        std::cerr << "Invalid second argument: " << argv[2]
                  << " only values of main or secondary are valid" << std::endl;
        return EXIT_FAILURE;
    }

    sensor::sensor_config config;
    if (!sensor::get_config(sensor_hostname, config)) {
        std::cerr << "Failed to get sensor config" << std::endl;
        return EXIT_FAILURE;
    }
    config.udp_dest = "239.201.201.201";
    std::shared_ptr<ouster::sensor::client> cli;
    if (main) {
        // This assumes a change to subscribe to IPADD_ANY if not specified.
        // Othewise you need to specify your own IP address such as
        // cli = sensor::mtp_init_client_main(sensor_hostname, config,
        // "192.168.1.11");
        cli = sensor::mtp_init_client_main(sensor_hostname, config);
    } else {
        cli = sensor::mtp_init_client_slave(sensor_hostname, config);
    }
    if (!cli) {
        std::cerr << "Failed to initialize sensor" << std::endl;
        return EXIT_FAILURE;
    }

    auto metadata = sensor::get_metadata(*cli);
    sensor::sensor_info info = sensor::parse_metadata(metadata);
    sensor::packet_format pf = sensor::get_format(info);
    auto packet_buf = std::make_unique<uint8_t[]>(UDP_BUF_SIZE);

    while (true) {
        sensor::client_state st = sensor::poll_client(*cli);

        if (st == sensor::TIMEOUT) {
            std::cerr << "Timed out" << std::endl;
            continue;
        }
        if (st & sensor::LIDAR_DATA) {
            if (sensor::read_lidar_packet(*cli, packet_buf.get(), pf)) {
                std::cerr << "Read Lidar Packet" << std::endl;
            }
        }

        if (st & sensor::IMU_DATA) {
            if (sensor::read_imu_packet(*cli, packet_buf.get(), pf)) {
                std::cerr << "Read IMU packet" << std::endl;
            }
        }
    }
}`

Samahu commented:
@ShepelIlya Sorry I got busy with many other stuff :tired_face: I'll try to catch on this PR tomorrow.

Here are few things that got my eyes on:
1. I think we need to hide the `mtp_init_client` method, as I perceived method is solely provided so it would be invoked by `mtp_init_client_main` and `mtp_init_client_secondary`. We don't want to confuse users, there are two versions of mtp_ methods (main and secondary), that's the API provides.
2. In response to [the thread](https://github.com/ouster-lidar/ouster_example/pull/480#discussion_r1105798908) I will test this again tomorrow.
3. In response to [the thread](https://github.com/ouster-lidar/ouster_example/pull/480#discussion_r1106318115) with regard to the new `CONFIG_GET_ACTIVE` flag, I am not 100% sure, but I still think it is not required and you can still achieve the same results for secondary mtp inits without having the flag added. Again, I won't be sure until I have a closer look and maybe prototype it to see. Since I may need to prototype it anyway till I find out, if I was successful at this without making the code more complicated I will probably push the code to a branch and open a merge request against this branch and let you review it (just to speed things up). 

Finally, I will double check that the code cross-compiles across all platforms that we support.

I plan to have this PR merged by the end of this week. Thanks for the help.

ShepelIlya commented:
> @ShepelIlya Sorry I got busy with many other stuff tired_face I'll try to catch on this PR tomorrow.
> 
> Here are few things that got my eyes on:
> 
>     1. I think we need to hide the `mtp_init_client` method, as I perceived the method is provided solely so it would be invoked by `mtp_init_client_main` and `mtp_init_client_secondary`. We don't want to confuse users, there are two versions of mtp_ that the client provides to users (_main and _secondary), that's what the API provides.
> 
>     2. In response to [the thread](https://github.com/ouster-lidar/ouster_example/pull/480#discussion_r1105798908) I will test this again tomorrow.
> 
>     3. In response to [the thread](https://github.com/ouster-lidar/ouster_example/pull/480#discussion_r1106318115) with regard to the new `CONFIG_GET_ACTIVE` flag, I am not 100% sure, but I still think it is not required and you can still achieve the same results for secondary mtp inits without having the flag added. Again, I won't be sure until I have a closer look and maybe prototype it to see. Since I may need to prototype it anyway till I find out, if I was successful at this without making the code more complicated I will probably push the code to a branch and open a merge request against this branch and let you review it (just to speed things up).
> 
> 
> Finally, I will double check that the code cross-compiles across all platforms that we support.
> 
> I plan to have this PR merged by the end of this week. Thanks for your help.

How do you want to hide `mtp_init_client`??  - I can do it tomorrow.

Samahu commented:
> > @ShepelIlya Sorry I got busy with many other stuff tired_face I'll try to catch on this PR tomorrow.
> > Here are few things that got my eyes on:
> > ```
> > 1. I think we need to hide the `mtp_init_client` method, as I perceived the method is provided solely so it would be invoked by `mtp_init_client_main` and `mtp_init_client_secondary`. We don't want to confuse users, there are two versions of mtp_ that the client provides to users (_main and _secondary), that's what the API provides.
> > 
> > 2. In response to [the thread](https://github.com/ouster-lidar/ouster_example/pull/480#discussion_r1105798908) I will test this again tomorrow.
> > 
> > 3. In response to [the thread](https://github.com/ouster-lidar/ouster_example/pull/480#discussion_r1106318115) with regard to the new `CONFIG_GET_ACTIVE` flag, I am not 100% sure, but I still think it is not required and you can still achieve the same results for secondary mtp inits without having the flag added. Again, I won't be sure until I have a closer look and maybe prototype it to see. Since I may need to prototype it anyway till I find out, if I was successful at this without making the code more complicated I will probably push the code to a branch and open a merge request against this branch and let you review it (just to speed things up).
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Finally, I will double check that the code cross-compiles across all platforms that we support.
> > I plan to have this PR merged by the end of this week. Thanks for your help.
> 
> How do you want to hide `mtp_init_client`?? - I can do it tomorrow.

Remove it from the header and keep the definition the cpp.

Samahu commented:
@ShepelIlya I discussed this with @MattC11 who is a member of our team, he suggested that it would be simpler to have one mtp_init_client method with a boolean that indicates whether this is a main or a secondary rather than providing two API, after reviewing the ouster-ros code and the changes you had to make I think I lean towards this opinion.

So the current suggestion is to maintain `mtp_init_client` method you had defined in `client.h` and add a new parameter `main` to it as follows:

```c++
/**
 * Connect to sensor and start listening for data via multicast.
 *
 * @param[in] hostname hostname or ip of the sensor. 
 * @param[in] config sensor config.
 * @param[in] mtp_dest_host multicast ip address where the sensor should send data.
 * @param[in] main a flag that indicates this is the main connection to the sensor in an mtp setup
 *  when this flag is set the config object will be used to configure the sensor, otherwise only the
 *  port values within the config object will be used and the rest will be ignored.
 *
 * @return pointer owning the resources associated with the connection.
 */
std::shared_ptr<client> mtp_init_client(const std::string& hostname,
                                        const sensor_config& config,
                                        const std::string& mtp_dest_host
                                        bool main);
```
This method would accomplish what the two method `mtp_init_client_main` and `mtp_init_client_secondary` are doing so we can remove them.

What we can do is pass this flag directly from the ouster-ros launch files, so we would have something like:
```bash
roslaunch ouster_ros sensor_mtp.launch \
   sensor_hostname:=$SENSOR_IP \ # always required
   udp_dest=$MULTICAST_IP \           # required arg for sensor_mtp.launch
   mtp_dest:=$CLIENT_IP \             # required arg for sensor_mtp.launch
   mtp_main:=true ..                  # we can set mtp_main default value in the launch file to false or leave it as a required argument


```
I think with that you don't need the `CONFIG_GET_ACTIVE` which based on my review it is only needed within the ouster-ros.

Sorry for the back and forth but it is a design issue, the whole idea is to avoid sensor init/reinit on every new secondary connection to the sensor.

Sorry for the back and forth but this is a design issue and the problem only surfaced to me after you added the `CONFIG_GET_ACTIVE`, I think if the launch command tells you readily that we are requesting a main vs secondary connection then you wouldn't need to do a lot work trying to guess what mode the user wants to run this session with.

Let me know what do you think?

Samahu commented:
Within ouster-ros/os_sensor_nodelet if `mtp_main` flag was set to false then we can bypass the `configure_sensor` method and proceed directly with `create_sensor_client() { auto cli = sensor::mtp_init_client(host, config, mtp_dest_host=mtp_host, mtp_main=false); ... return cli }`.

Samahu commented:
I did run your code by our cross-platform compilation checks and it didn't report any problems.

ShepelIlya commented:
> @ShepelIlya I discussed this with @MattC11 who is a member of our team, he suggested that it would be simpler to have one `mtp_init_client` method with a **boolean** flag that indicates whether this is a main or a secondary rather than providing two separate APIs, after reviewing the _ouster-ros_ code and the changes you had to make I think I lean towards his opinion.
> 
> So the current suggestion is to maintain `mtp_init_client` method you had defined in `client.h` and add a new parameter `main` to it as follows:
> 
> ```c++
> /**
>  * Connect to sensor and start listening for data via multicast.
>  *
>  * @param[in] hostname hostname or ip of the sensor. 
>  * @param[in] config sensor config.
>  * @param[in] mtp_dest_host multicast ip address where the sensor should send data.
>  * @param[in] main a flag that indicates this is the main connection to the sensor in an mtp setup
>  *  when this flag is set the config object will be used to configure the sensor, otherwise only the
>  *  port values within the config object will be used and the rest will be ignored.
>  *
>  * @return pointer owning the resources associated with the connection.
>  */
> std::shared_ptr<client> mtp_init_client(const std::string& hostname,
>                                         const sensor_config& config,
>                                         const std::string& mtp_dest_host
>                                         bool main,
>                                         int timeout_sec = 60);
> ```
> 
> This method would accomplish what the two method `mtp_init_client_main` and `mtp_init_client_secondary` are doing so we can remove them.
> 
> What we can do is pass this flag directly from the ouster-ros launch files, so we would have something like:
> 
> ```shell
> roslaunch ouster_ros sensor_mtp.launch \
>    sensor_hostname:=$SENSOR_IP \ # always required
>    udp_dest=$MULTICAST_IP \           # required arg for sensor_mtp.launch
>    mtp_dest:=$CLIENT_IP \             # required arg for sensor_mtp.launch
>    mtp_main:=true ..                  # we can set mtp_main default value in the launch file to false or leave it as a required argument
> ```
> 
> I think with that you don't need the `CONFIG_GET_ACTIVE` which based on my review it is only needed within the ouster-ros.
> 
> Sorry for the back and forth but it is a design issue, the whole idea is to avoid sensor init/reinit on every new secondary connection to the sensor.
> 
> Sorry for the back and forth but this is a design issue and the problem only surfaced to me after you added the `CONFIG_GET_ACTIVE`, I think if the launch command tells you readily that we are requesting a main vs secondary connection then you wouldn't need to do a lot work trying to guess what mode the user wants to run this session with.
> 
> Let me know what do you think?

Then I can add mtp_main flag directly - its ok for me, my biggest motivation for CONFIG_GET ACTIVE was to keep user from configuring based on mtp settings. If its ok - to give him choice, than adding flag to launch is simpler.

I'll do it later today than

ShepelIlya commented:
@Samahu i think i did it right now:

[sdk](https://github.com/ouster-lidar/ouster_example/pull/480/commits/fcd8cda266864310e3d6f5fc20fcbcf6e1bb08c5):
1. CONFIG_GET_ACTIVE deleted
2.  mtp_init_client_main and mtp_init_client_secondary merged to mtp_initi_client with flag main

[driver](https://github.com/ouster-lidar/ouster-ros/pull/57/commits/ef4297a022719a343063df9be4048d0d4af24906):
1. corresponding `mtp_init_client` update
2. added `bool mtp_main` flag
3. deleted sensor_mtp_slave.launch
4. flag mtp_main added to sensor_mtp.launch


Check it please.

Samahu commented:
taking a look right now

Samahu commented on ouster_client/include/ouster/client.h:
```diff
@@ -97,6 +97,23 @@ std::shared_ptr<client> init_client(const std::string& hostname,
                                     timestamp_mode ts_mode = TIME_FROM_UNSPEC,
                                     int lidar_port = 0, int imu_port = 0,
                                     int timeout_sec = 60);
+
+/**
+ * Connect to and configure the sensor and start listening for data via multicast.
+ *
+ * @param[in] hostname hostname or ip of the sensor. 
+ * @param[in] config sensor config to set on sensor.
+ * @param[in] mtp_dest_host multicast ip address where the sensor should send data.
+ * @param[in] timeout_sec how long to wait for the sensor to initialize.
+ *
+ * @return pointer owning the resources associated with the connection.
+ */
+std::shared_ptr<client> mtp_init_client(const std::string& hostname,                                    
+                                             const sensor_config& config,
+                                             const std::string& mtp_dest_host,
+                                             bool main,
+                                             int timeout_sec = 60);
+
```
documentation + formatting

```suggestion

/**
 * [BETA] Connect to and configure the sensor and start listening for data via
 * multicast.
 *
 * @param[in] hostname hostname or ip of the sensor.
 * @param[in] config sensor config to set on sensor.
 * @param[in] mtp_dest_host multicast ip address where the sensor should send
 * data.
 * @param[in] main a flag that indicates this is the main connection to the
 * sensor in an multicast setup.
 * @param[in] timeout_sec how long to wait for the sensor to initialize.
 *
 * @return pointer owning the resources associated with the connection.
 *
 * @remarks when main flag is set the config object will be used to configure
 * the sensor, otherwise only the port values within the config object will be
 * used and the rest will be ignored.
 */
std::shared_ptr<client> mtp_init_client(const std::string& hostname,
                                        const sensor_config& config,
                                        const std::string& mtp_dest_host,
                                        bool main, int timeout_sec = 60);

```

Samahu commented on ouster_client/src/client.cpp:
```diff
@@ -154,6 +154,95 @@ SOCKET udp_data_socket(int port) {
+                mreq.imr_multiaddr.s_addr = inet_addr(udp_dest_host.c_str());
+                if (!mtp_dest_host.empty()) {
+                    mreq.imr_interface.s_addr = inet_addr(mtp_dest_host.c_str());
+                } else {
+                    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
+                }
+
+                if (setsockopt(sock_fd, IPPROTO_IP, IP_ADD_MEMBERSHIP, (char*)&mreq,
+                            sizeof(mreq))) {
+                    logger().warn("mtp setsockopt(): {}", impl::socket_get_error());
+                    impl::socket_close(sock_fd);
+                    continue;
+                }
+            }
+
+            // join to multicast group succeeded; set some options and return
+            if (impl::socket_set_non_blocking(sock_fd)) {
+                logger().warn("mtp fcntl(): {}", impl::socket_get_error());
+                impl::socket_close(sock_fd);
+                continue;
+            }
+
+            if (setsockopt(sock_fd, SOL_SOCKET, SO_RCVBUF, (char*)&RCVBUF_SIZE,
+                           sizeof(RCVBUF_SIZE))) {
+                logger().warn("mtp setsockopt(): {}", impl::socket_get_error());
+                impl::socket_close(sock_fd);
+                continue;
+            }
+
+            freeaddrinfo(info_start);
+            return sock_fd;
+        }
+    }
+
+    // could not bind() a MTP server socket
+    freeaddrinfo(info_start);
+    logger().error("failed to bind mtp socket");
+    return SOCKET_ERROR;
+}
+
```
[formatting]: remove many white spaces
```suggestion
SOCKET mtp_data_socket(int port, const std::string& udp_dest_host = "",
                       const std::string& mtp_dest_host = "") {
    struct addrinfo hints, *info_start, *ai;

    memset(&hints, 0, sizeof hints);
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_DGRAM;
    hints.ai_flags = AI_PASSIVE;

    auto port_s = std::to_string(port);

    int ret = getaddrinfo(NULL, port_s.c_str(), &hints, &info_start);
    if (ret != 0) {
        logger().error("mtp getaddrinfo(): {}", gai_strerror(ret));
        return SOCKET_ERROR;
    }
    if (info_start == NULL) {
        logger().error("mtp getaddrinfo(): empty result");
        return SOCKET_ERROR;
    }

    for (auto preferred_af : {AF_INET}) {  // TODO test with AF_INET6
        for (ai = info_start; ai != NULL; ai = ai->ai_next) {
            if (ai->ai_family != preferred_af) continue;

            // choose first addrinfo where bind() succeeds
            SOCKET sock_fd =
                socket(ai->ai_family, ai->ai_socktype, ai->ai_protocol);
            if (!impl::socket_valid(sock_fd)) {
                logger().warn("mtp socket(): {}", impl::socket_get_error());
                continue;
            }

            if (impl::socket_set_reuse(sock_fd)) {
                logger().warn("mtp socket_set_reuse(): {}",
                              impl::socket_get_error());
            }

            if (::bind(sock_fd, ai->ai_addr, (socklen_t)ai->ai_addrlen)) {
                logger().warn("mtp bind(): {}", impl::socket_get_error());
                impl::socket_close(sock_fd);
                continue;
            }

            // bind() succeeded; join to multicast group on with preferred
            // address connect only if addresses are not empty
            if (!udp_dest_host.empty()) {
                ip_mreq mreq;
                mreq.imr_multiaddr.s_addr = inet_addr(udp_dest_host.c_str());
                if (!mtp_dest_host.empty()) {
                    mreq.imr_interface.s_addr =
                        inet_addr(mtp_dest_host.c_str());
                } else {
                    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
                }

                if (setsockopt(sock_fd, IPPROTO_IP, IP_ADD_MEMBERSHIP,
                               (char*)&mreq, sizeof(mreq))) {
                    logger().warn("mtp setsockopt(): {}",
                                  impl::socket_get_error());
                    impl::socket_close(sock_fd);
                    continue;
                }
            }

            // join to multicast group succeeded; set some options and return
            if (impl::socket_set_non_blocking(sock_fd)) {
                logger().warn("mtp fcntl(): {}", impl::socket_get_error());
                impl::socket_close(sock_fd);
                continue;
            }

            if (setsockopt(sock_fd, SOL_SOCKET, SO_RCVBUF, (char*)&RCVBUF_SIZE,
                           sizeof(RCVBUF_SIZE))) {
                logger().warn("mtp setsockopt(): {}", impl::socket_get_error());
                impl::socket_close(sock_fd);
                continue;
            }

            freeaddrinfo(info_start);
            return sock_fd;
        }
    }

    // could not bind() a MTP server socket
    freeaddrinfo(info_start);
    logger().error("failed to bind mtp socket");
    return SOCKET_ERROR;
}

```

Samahu commented on ouster_client/src/client.cpp:
```diff
@@ -380,6 +469,53 @@ std::shared_ptr<client> init_client(const std::string& hostname,
+    logger().info("initializing sensor client: {} with ports: {}/{}, multicast group: {}",
+                  hostname, config.udp_port_lidar.value(), config.udp_port_imu.value(),
+                  config.udp_dest.value());
+
+    auto cli = std::make_shared<client>();
+    cli->hostname = hostname;
+
+    cli->lidar_fd = mtp_data_socket(config.udp_port_lidar.value(), config.udp_dest.value(), mtp_dest_host);
+    cli->imu_fd = mtp_data_socket(config.udp_port_imu.value()); // no need to join multicast group second time
+
+    if (!impl::socket_valid(cli->lidar_fd) || !impl::socket_valid(cli->imu_fd))
+        return std::shared_ptr<client>();
+
+    if (main) {        
+        auto lidar_port = get_sock_port(cli->lidar_fd);
+        auto imu_port = get_sock_port(cli->imu_fd);
+
+        sensor_config config_copy{config};
+        try {
+            uint8_t config_flags = 0;        
+            if (lidar_port) config_copy.udp_port_lidar = lidar_port;
+            if (imu_port) config_copy.udp_port_imu = imu_port;
+            config_copy.operating_mode = OPERATING_NORMAL;
+            set_config(hostname, config_copy, config_flags);
+
+            // will block until no longer INITIALIZING
+            cli->meta = collect_metadata(hostname, timeout_sec);
+            // check for sensor error states
+            auto status = cli->meta["sensor_info"]["status"].asString();
+            if (status == "ERROR" || status == "UNCONFIGURED")
+                return std::shared_ptr<client>();
+        } catch (const std::runtime_error& e) {
+            // log error message
+            logger().error("init_client(): {}", e.what());
+            return std::shared_ptr<client>();
+        }
+    }
+
+    return cli;   
+}
```
[required] function name +formatting
```suggestion
std::shared_ptr<client> mtp_init_client(const std::string& hostname,
                                        const sensor_config& config,
                                        const std::string& mtp_dest_host,
                                        bool main, int timeout_sec) {
    logger().info(
        "initializing sensor client: {} with ports: {}/{}, multicast group: {}",
        hostname, config.udp_port_lidar.value(), config.udp_port_imu.value(),
        config.udp_dest.value());

    auto cli = std::make_shared<client>();
    cli->hostname = hostname;

    cli->lidar_fd = mtp_data_socket(config.udp_port_lidar.value(),
                                    config.udp_dest.value(), mtp_dest_host);
    cli->imu_fd = mtp_data_socket(
        config.udp_port_imu
            .value());  // no need to join multicast group second time

    if (!impl::socket_valid(cli->lidar_fd) || !impl::socket_valid(cli->imu_fd))
        return std::shared_ptr<client>();

    if (main) {
        auto lidar_port = get_sock_port(cli->lidar_fd);
        auto imu_port = get_sock_port(cli->imu_fd);

        sensor_config config_copy{config};
        try {
            uint8_t config_flags = 0;
            if (lidar_port) config_copy.udp_port_lidar = lidar_port;
            if (imu_port) config_copy.udp_port_imu = imu_port;
            config_copy.operating_mode = OPERATING_NORMAL;
            set_config(hostname, config_copy, config_flags);

            // will block until no longer INITIALIZING
            cli->meta = collect_metadata(hostname, timeout_sec);
            // check for sensor error states
            auto status = cli->meta["sensor_info"]["status"].asString();
            if (status == "ERROR" || status == "UNCONFIGURED")
                return std::shared_ptr<client>();
        } catch (const std::runtime_error& e) {
            // log error message
            logger().error("init_client(): {}", e.what());
            return std::shared_ptr<client>();
        }
    }

    return cli;
}
```

Samahu commented on ouster_client/include/ouster/client.h:
```diff
@@ -213,5 +230,14 @@ int get_lidar_port(client& cli);
+ *
+ * @return true if addr is in multicast range.
+ */
+bool in_multicast(const char* addr);
```
We largely depend on `std::string` so I would rather use it than raw string
```suggestion
bool in_multicast(const std::string& addr);
```
Within the function body we can call addr.c_str()

ShepelIlya replied:
Fixed in [450aefc](https://github.com/ouster-lidar/ouster_example/pull/480/commits/450aefce39c22c60870b515ae91e9f55bf4446d1)

Samahu reviewed (commented):
Overall looks good to me, I made few minor comments:
- [ ] documentation
- [ ] formatting
- [ ] function name that needs to be renamed
- [ ] Additionally, we need to add a new entry to the [CHANGELOG](
https://github.com/ouster-lidar/ouster_example/blob/master/CHANGELOG.rst), it should say something like:
```
[unreleased]
============

ouster_client
-------------
- Added a new method ``mtp_init_client`` to init the client with multicast support (experimental).
```
We need to mention that this is experimental since we didn't test it thoroughly. 
- [ ] this optional but @MattC11 suggested that we could add [the code sample he wrote using your mtp method](https://github.com/ouster-lidar/ouster_example/pull/480#issuecomment-1430098559) to our [C++ examples](https://github.com/ouster-lidar/ouster_example/tree/master/examples). I think this would be very helpful to users interested in using the method without ROS. but I understand this wasn't originally part of your PR and we can add the example later ourselves. I am just giving you the option in case you wanted to add it as part of this PR.

MattC11 commented:
@ShepelIlya Just to say thanks for all your work on this. We realize there have been a lot of requests for changes.

ShepelIlya commented:
Sorry for formatting and name mismatch in last commit, i was in a rush ;(  
I added all your suggestions for now. After merge of this PR it will be much easier for me to keep our fork (with our custom features) actual. Looking forward to discussing them (may be i'll create some issues, just to know if you are intrested in)

Samahu commented:
> Sorry for formatting and name mismatch in last commit, i was in a rush ;( I added all your suggestions for now. After merge of this PR it will be much easier for me to keep our fork (with our custom features) actual. Looking forward to discussing them (may be i'll create some issues, just to know if you are intrested in)

No worries, I am doing final tests, we can discuss your suggestions later on.

MattC11 commented on examples/mtp_client_example.cpp:
```diff
@@ -0,0 +1,83 @@
+        // "192.168.1.11");
+        cli = sensor::mtp_init_client_main(sensor_hostname, config);
+    } else {
+        cli = sensor::mtp_init_client_secondary(sensor_hostname, config);
```
My example use the old function names, you should remove the if and change to (I think):

mtp_init_client(sensor_hostname, config, "", main)

Samahu replied:
Yeah I noticed that when building. I was going to write a suggestion like this:

```cpp
    // In case of main true, This assumes a change to subscribe to IPADDR_ANY if not specified.
    // Othewise you need to specify your own IP address such as
    // cli = sensor::mtp_init_client_main(sensor_hostname, config,
    // "192.168.1.11");

    std::shared_ptr<ouster::sensor::client> cli = sensor::mtp_init_client(
            sensor_hostname, config, "", main);

    if (!cli) {
        std::cerr << "Failed to initialize sensor" << std::endl;
        return EXIT_FAILURE;
    }
```

but I haven't tested it yet.

MattC11 replied:
Without the comment above which can be remove now since it default to ANY yes that looks right and it compiles.

Samahu replied:
@ShepelIlya I will make a suggestion on the entire example shortly that you can simply approve

Samahu commented on examples/mtp_client_example.cpp:
```diff
@@ -0,0 +1,83 @@
+    std::shared_ptr<ouster::sensor::client> cli;
+    if (main) {
+        // This assumes a change to subscribe to IPADDR_ANY if not specified.
+        // Othewise you need to specify your own IP address such as
+        // cli = sensor::mtp_init_client_main(sensor_hostname, config,
+        // "192.168.1.11");
+        cli = sensor::mtp_init_client_main(sensor_hostname, config);
+    } else {
+        cli = sensor::mtp_init_client_secondary(sensor_hostname, config);
+    }
+    if (!cli) {
+        std::cerr << "Failed to initialize sensor" << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    auto metadata = sensor::get_metadata(*cli);
+    sensor::sensor_info info = sensor::parse_metadata(metadata);
+    sensor::packet_format pf = sensor::get_format(info);
+    auto packet_buf = std::make_unique<uint8_t[]>(UDP_BUF_SIZE);
+
+    while (true) {
+        sensor::client_state st = sensor::poll_client(*cli);
+
+        if (st == sensor::TIMEOUT) {
+            std::cerr << "Timed out" << std::endl;
+            continue;
+        }
+        if (st & sensor::LIDAR_DATA) {
+            if (sensor::read_lidar_packet(*cli, packet_buf.get(), pf)) {
+                std::cerr << "Read Lidar Packet" << std::endl;
+            }
+        }
+
+        if (st & sensor::IMU_DATA) {
+            if (sensor::read_imu_packet(*cli, packet_buf.get(), pf)) {
+                std::cerr << "Read IMU packet" << std::endl;
+            }
+        }
+    }
+}
```
```suggestion
/**
 * Copyright (c) 2023, Ouster, Inc.
 * All rights reserved.
 */

#include <iostream>

#include "ouster/client.h"

using namespace ouster;

const size_t UDP_BUF_SIZE = 65536;

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "\n\nUsage: <sensor_hostname> <main|secondary>"
                  << std::endl;

        return argc == 1 ? EXIT_SUCCESS : EXIT_FAILURE;
    }

    const std::string sensor_hostname = argv[1];

    bool main = false;

    if (std::string(argv[2]) == "main") {
        main = true;
    } else if (std::string(argv[2]) == "secondary") {
        main = false;
    } else {
        std::cerr << "Invalid second argument: " << argv[2]
                  << " only values of main or secondary are valid" << std::endl;
        return EXIT_FAILURE;
    }

    sensor::sensor_config config;
    if (!sensor::get_config(sensor_hostname, config)) {
        std::cerr << "Failed to get sensor config" << std::endl;
        return EXIT_FAILURE;
    }
    config.udp_dest = "239.201.201.201";
    if (sensor::in_multicast(config.udp_dest.value())) {
        std::cerr << "In multicast" << std::endl;
    } else {
        std::cerr << "Not a multicast address" << std::endl;
        return -1;
    }

    std::shared_ptr<ouster::sensor::client> cli =
        sensor::mtp_init_client(sensor_hostname, config, "", main);

    if (!cli) {
        std::cerr << "Failed to initialize sensor" << std::endl;
        return EXIT_FAILURE;
    }

    auto metadata = sensor::get_metadata(*cli);
    sensor::sensor_info info = sensor::parse_metadata(metadata);
    sensor::packet_format pf = sensor::get_format(info);
    auto packet_buf = std::make_unique<uint8_t[]>(UDP_BUF_SIZE);

    bool done = false;
    while (!done) {
        sensor::client_state state = sensor::poll_client(*cli);
        if (state == sensor::EXIT) {
            std::cerr << "caught signal, exiting" << std::endl;
            done = true;
        }
        if (state == sensor::TIMEOUT) {
            std::cerr << "Timed out" << std::endl;
            continue;
        }
        if (state & sensor::LIDAR_DATA) {
            if (sensor::read_lidar_packet(*cli, packet_buf.get(), pf)) {
                std::cerr << "Read Lidar Packet" << std::endl;
            }
        }
        if (state & sensor::IMU_DATA) {
            if (sensor::read_imu_packet(*cli, packet_buf.get(), pf)) {
                std::cerr << "Read IMU packet" << std::endl;
            }
        }
    }
}
```

compiles and tested

Samahu reviewed (approved):
@ShepelIlya I committed the fixes for the `mtp_client_example.cpp`

With that I think everything is in shape and I have this PR approved :tada: :tada: :tada:  (I will follow up on the ouster-ros PR soon)

also adding @kairenw for any further feedback before this get's merged

kairenw commented on CHANGELOG.rst:
```diff
@@ -2,6 +2,14 @@
 Changelog
 =========
 
+[unreleased]
+============
+
+ouster_client
+-------------
+- Added a new method ``mtp_init_client`` to init the client with multicast support (experimental).
+
+
```
:tada:

kairenw commented on ouster_client/include/ouster/client.h:
```diff
@@ -97,6 +97,30 @@ std::shared_ptr<client> init_client(const std::string& hostname,
+
+/**
+ * [BETA] Connect to and configure the sensor and start listening for data via
+ * multicast.
```
can we specify here that currently this is supported/tested only on Linux (though it will build on Mac/Windows)

kairenw replied:
Discussed with @Samahu - we will test it later on our side. Since it's marked [BETA], we think it's covered for now.
Will resolve conversation.

Samahu commented:
I would rather try to test it, unfortunately I don’t have a windows machine
ready but I can try to setup one later today..

On Fri, Feb 17, 2023 at 1:18 PM kairenw ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In ouster_client/include/ouster/client.h
> <https://github.com/ouster-lidar/ouster_example/pull/480#discussion_r1110377651>
> :
>
> > @@ -97,6 +97,30 @@ std::shared_ptr<client> init_client(const std::string& hostname,
>                                      timestamp_mode ts_mode = TIME_FROM_UNSPEC,
>                                      int lidar_port = 0, int imu_port = 0,
>                                      int timeout_sec = 60);
> +
> +/**
> + * [BETA] Connect to and configure the sensor and start listening for data via
> + * multicast.
>
> can we specify here that currently this is supported/tested only on Linux
> (though it will build on Mac/Windows)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ouster-lidar/ouster_example/pull/480#pullrequestreview-1304218168>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAET6UOBMXEYJ727YYO75CTWX7TKXANCNFSM6AAAAAAUJW6Q3U>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
*Ussama Naal*
San Francisco, CA
<http://ouster.com/>

kairenw commented:
:tada: :tada: :tada: :tada: :tada: :exclamation:
